solutions/cf-1359C.mdx

```diff
@@ -0,0 +1,114 @@
+- Case 1: pouring an equal number of cups of hot and cold water is equivalent to pouring one cup of each. Otherwise,
+- Case 2: there must be exactly one more cup of hot water poured than cold water poured. Thus, there will be $c_i + 1$ cups of hot water and $c_i$ cups of cold water for some $c \geq 0$.
+- It can be proven, using induction, that the average barrel temperatures when $c$ is odd is a monotonically decreasing function. Thus, we can binary search on the maximum number of odd cups $c$ which gives a temperature of at least $t$ by checking the odd integers around it.
+
+The answer will be either be the barrel with $2$, $c$, or $c + 2$ cups.
+
+**Time Complexity:** $\mathcal{O}(\log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+double hot_water, cold_water, desired_temp;
+
+/*
+   * This function checks whether pouring cupsPoured amount of cups achieves an average
+   * barrel temperature greater than or equal to desired_temp. Technically, this represents
+   * 2 * cupsPoured + 1 cups
+*/
+
+bool works (int cupsPoured) {
+	double totalBarrelTemperature = (hot_water * (cupsPoured + 1) + cold_water * (cupsPoured));
+	double totalCupsPoured = (double) (2 * cupsPoured + 1.0);
+	double avgBarrelTemperature = totalBarrelTemperature / totalCupsPoured;
+	return avgBarrelTemperature >= desired_temp;
+}
+
+void solve () {
+	/* 
+	   * maximumCups represents the the maximum number of cups such
+	   * that the barrel temperature is greater than desired_temp.
+	*/
+	cin >> hot_water >> cold_water >> desired_temp;
+	// Binary search on the answer for maximumCups.
+	int left = 0, right = 1e6 + 1, maximumCups = 0;
+
+	while (left <= right) {
+		int mid = left + (right - left) / 2;
+		if (works(mid)) {
+			maximumCups = mid;
+			left = mid + 1;
+		} else {
+			right = mid - 1;
+		}
+	}
+
+	// If it is impossible to find a temperature close to desired_temp, then print 2
+	if (maximumCups == 1e6 + 1) {
+		cout << 2 << endl;
+	} else {
+		/*
+		   * The solution must lie between barrel1 and barrel2 if the total number of cups
+		   * is odd. barrel1 represents the barrel with an average temperature slightly
+		   * greater than desired_temp. barrel2 represent the barrel with an average
+		   * temperature slightly less than desired_temp. These two barrels will be the
+		   * closest to desired_temp.
+		*/
+		double totalBarrel1 = (hot_water * (maximumCups + 1) + cold_water * maximumCups);
+		double totalBarrel2 = (hot_water * (maximumCups + 2) + cold_water * (maximumCups + 1));
+		double barrel1Cups = (double)(2 * maximumCups + 1.0);
+		double barrel2Cups = (double)(2 * maximumCups + 3.0);
+		double avgBarrel1 = totalBarrel1 / barrel1Cups;
+		double avgBarrel2 = totalBarrel2 / barrel2Cups;
+		double hot_cold_avg = (hot_water + cold_water) / 2.0;
+
+		double minAbsDifference = min({abs(desired_temp - avgBarrel1),
+										abs(desired_temp - avgBarrel2),
+										abs(desired_temp - hot_cold_avg)});
```

Comment:
i don't think this formatting is correct

Reply:
i think this was me... dw, it is correct - github just makes tabs larger which makes the spacing seem off
![image](https://user-images.githubusercontent.com/63619830/120212734-31150780-c1e7-11eb-86d4-ee9ec65f4d61.png)